File: idea-plugin/src/main/kotlin/com/gitai/commit/OpenAiCompatibleClient.kt
package com.gitai.commit

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import java.io.IOException
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.util.stream.Stream

data class OpenAiChatMessage(
    val role: String,
    val content: String
)

data class OpenAiChatCompletionRequest(
    val model: String,
    val messages: List<OpenAiChatMessage>,
    val temperature: Double = 0.0,
    val stream: Boolean = false
)

class OpenAiCompatibleClient(
    private val baseUrl: String,
    private val apiKey: String = "",
    private val httpClient: HttpClient = HttpClient.newHttpClient(),
    private val objectMapper: ObjectMapper = ObjectMapper()
) : ModelProvider {
    override val id: String = "openai-compatible"

    override fun generate(model: String, prompt: String): String {
        return generateStream(model, prompt) { }
    }

    override fun generateStream(model: String, prompt: String, onChunk: (String) -> Unit): String {
        val body = objectMapper.writeValueAsString(
            OpenAiChatCompletionRequest(
                model = model,
                messages = listOf(
                    OpenAiChatMessage("system", "只输出一句中文提交信息，不要输出 JSON、不要输出英文提交信息、不要解释、不要正文。"),
                    OpenAiChatMessage("user", prompt)
                ),
                temperature = 0.0,
                stream = true
            )
        )
        val requestBuilder = HttpRequest.newBuilder()
            .uri(URI.create("${baseUrl.trimEnd('/')}/chat/completions"))
            .header("Content-Type", "application/json")
        if (apiKey.isNotBlank()) {
            requestBuilder.header("Authorization", "Bearer $apiKey")
        }
        val request = requestBuilder
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .build()
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofLines())
        if (response.statusCode() >= 400) {
            throw IOException("OpenAI-compatible provider returned HTTP ${response.statusCode()}: ${response.body()}")
        }

        val raw = StringBuilder()
        response.body().use { lines ->
            lines.forEach { line ->
                val delta = extractOpenAiCompatibleDelta(line)
                if (delta.isNotEmpty()) {
                    raw.append(delta)
                    onChunk(delta)
                }
            }
        }
        return raw.toString().trim()
    }
}

private fun extractOpenAiCompatibleDelta(line: String): String {
    val trimmed = line.trim()
    if (!trimmed.startsWith("data:")) {
        return ""
    }

    val payload = trimmed.removePrefix("data:").trim()
    if (payload.isEmpty() || payload == "[DONE]") {
        return ""
    }

    return try {
        val json = ObjectMapper().readTree(payload)
        val delta = json.path("choices")
            .path(0)
            .path("delta")
            .path("content")
            .asText("")
        if (delta.isNotEmpty()) {
            delta
        } else {
            json.path("choices")
                .path(0)
                .path("message")
                .path("content")
                .asText("")
        }
    } catch (_: Exception) {
        ""
    }
}

File: vscode-plugin/src/providers/openAiCompatibleClient.ts
import { ModelProvider, ModelProviderFactory, GitAiSettings, PROVIDER_BASE_URLS } from './modelProvider';

export function extractOpenAiCompatibleDelta(line: string): string {
    const trimmed = line.trim();
    if (!trimmed.startsWith('data:')) {
        return '';
    }

    const payload = trimmed.replace(/^data:\s*/, '');
    if (!payload || payload === '[DONE]') {
        return '';
    }

    try {
        const json = JSON.parse(payload) as {
            choices?: Array<{
                delta?: { content?: string };
                message?: { content?: string };
            }>;
        };
        return (json.choices?.[0]?.delta?.content ?? json.choices?.[0]?.message?.content ?? '');
    } catch {
        return '';
    }
}

async function* readResponseLines(response: Response): AsyncGenerator<string> {
    const reader = response.body?.getReader();
    if (!reader) {
        return;
    }

    const decoder = new TextDecoder();
    let buffer = '';

    while (true) {
        const { done, value } = await reader.read();
        if (done) {
            break;
        }

        buffer += decoder.decode(value, { stream: true });

        let newlineIndex = buffer.indexOf('\n');
        while (newlineIndex >= 0) {
            const line = buffer.slice(0, newlineIndex).replace(/\r$/, '');
            buffer = buffer.slice(newlineIndex + 1);
            yield line;
            newlineIndex = buffer.indexOf('\n');
        }
    }

    const tail = `${buffer}${decoder.decode()}`.trimEnd();
    if (tail) {
        yield tail;
    }
}

/**
 * OpenAI-compatible client matching the Kotlin `OpenAiCompatibleClient`.
 * Calls any OpenAI-compatible /chat/completions endpoint.
 */
export class OpenAiCompatibleClient implements ModelProvider {
    readonly id: string = 'openai-compatible';

    constructor(
        private baseUrl: string,
        private apiKey: string = '',
    ) {}

    async generate(model: string, prompt: string): Promise<string> {
        let raw = '';
        await this.generateStream(model, prompt, chunk => {
            raw += chunk;
        });
        return raw.trim();
    }

    async generateStream(model: string, prompt: string, onChunk: (chunk: string) => void): Promise<string> {
        const url = `${this.baseUrl.replace(/\/+$/, '')}/chat/completions`;
        const body = JSON.stringify({
            model,
            messages: [
                {
                    role: 'system',
                    content: '只输出一句中文提交信息，不要输出 JSON、不要输出英文提交信息、不要解释、不要正文。',
                },
                {
                    role: 'user',
                    content: prompt,
                },
            ],
            stream: true,
        });

        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
        };
        if (this.apiKey) {
            headers['Authorization'] = `Bearer ${this.apiKey}`;
        }

        const response = await fetch(url, {
            method: 'POST',
            headers,
            body,
        });

        if (!response.ok) {
            const text = await response.text();
            throw new Error(`OpenAI-compatible provider returned HTTP ${response.status}: ${text}`);
        }

        let raw = '';
        for await (const line of readResponseLines(response)) {
            const delta = extractOpenAiCompatibleDelta(line);
            if (!delta) {
                continue;
            }
            raw += delta;
            onChunk(delta);
        }

        return raw.trim();
    }
}

export class OpenAiCompatibleProviderFactory implements ModelProviderFactory {
    readonly id: string = 'openai-compatible';
    readonly defaultModel: string = '';

    create(settings: GitAiSettings): ModelProvider {
        return new OpenAiCompatibleClient(
            settings.openAiCompatibleBaseUrl,
            settings.openAiCompatibleApiKey,
        );
    }
}

// --- Cloud provider factories (matching Kotlin CloudProviderFactories.kt) ---

export class DeepSeekProviderFactory implements ModelProviderFactory {
    readonly id: string = 'deepseek';
    readonly defaultModel: string = 'deepseek-v4-flash';

    create(settings: GitAiSettings): ModelProvider {
        return new OpenAiCompatibleClient(
            settings.deepSeekBaseUrl || PROVIDER_BASE_URLS.DEEPSEEK,
            settings.openAiCompatibleApiKey,
        );
    }
}

export class AliyunProviderFactory implements ModelProviderFactory {
    readonly id: string = 'aliyun';
    readonly defaultModel: string = 'qwen-plus';

    create(settings: GitAiSettings): ModelProvider {
        return new OpenAiCompatibleClient(
            settings.aliyunBaseUrl || PROVIDER_BASE_URLS.ALIYUN,
            settings.openAiCompatibleApiKey,
        );
    }
}

export class MiniMaxProviderFactory implements ModelProviderFactory {
    readonly id: string = 'minimax';
    readonly defaultModel: string = 'MiniMax-Text-01';

    create(settings: GitAiSettings): ModelProvider {
        return new OpenAiCompatibleClient(
            settings.miniMaxBaseUrl || PROVIDER_BASE_URLS.MINIMAX,
            settings.openAiCompatibleApiKey,
        );
    }
}

export class KimiProviderFactory implements ModelProviderFactory {
    readonly id: string = 'kimi';
    readonly defaultModel: string = 'kimi-k2.5';

    create(settings: GitAiSettings): ModelProvider {
        return new OpenAiCompatibleClient(
            settings.kimiBaseUrl || PROVIDER_BASE_URLS.KIMI,
            settings.openAiCompatibleApiKey,
        );
    }
}

export class GlmProviderFactory implements ModelProviderFactory {
    readonly id: string = 'glm';
    readonly defaultModel: string = 'glm-4-flash';

    create(settings: GitAiSettings): ModelProvider {
        return new OpenAiCompatibleClient(
            settings.glmBaseUrl || PROVIDER_BASE_URLS.GLM,
            settings.openAiCompatibleApiKey,
        );
    }
}
